File: linked_list/reverse_linked_list2/solution.go
package main

import "fmt"

type ListNode struct {
	Val int
	Next *ListNode
}

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	if head == nil || left == right{
		return head
	}

	dummy := &ListNode{0,head}
	prev := dummy

	// move prev to the node before "left" position
	for i := 0; i < left; i++ {
		prev = prev.Next
	}

	curr := prev.Next
	var nextNode *ListNode

	// Reverse the sublist between "left" and "right"
	for i := left; i < right; i++ {
    nextNode = curr.Next
    curr.Next = nextNode.Next
    nextNode.Next = prev.Next
    prev.Next = nextNode
  }

  return dummy.Next
}

// 🔹 Test Case
func printList(head *ListNode) {
    for head != nil {
        fmt.Print(head.Val)
        if head.Next != nil {
            fmt.Print(" → ")
        }
        head = head.Next
    }
    fmt.Println()
}

func main() {
    head := &ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, &ListNode{5, nil}}}}}

    new_head := reverseBetween(head, 2, 4)
    printList(new_head) // Expected Output: 1 → 4 → 3 → 2 → 5
}

File: linked_list/reverse_linked_list2/solution.py
import sys
sys.stdout.reconfigure(encoding='utf-8')


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseBetween(self, head: ListNode, left: int, right: int) -> ListNode:
        if not head or left == right:
            return head

        dummy = ListNode(0)
        dummy.next = head
        prev = dummy

        # Move prev to node before left position
        for _ in range(left - 1):
            prev = prev.next

        curr = prev.next
        nextNode = None

        # Reverse sublist from left to right
        for _ in range(right - left):
            nextNode = curr.next
            curr.next = nextNode.next
            nextNode.next = prev.next
            prev.next = nextNode

        return dummy.next

# 🔹 Test Case
def printList(head):
    while head:
        print(head.val, end=" → " if head.next else "\n")
        head = head.next

head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
solution = Solution()
new_head = solution.reverseBetween(head, 2, 4)
printList(new_head)  # Expected Output: 1 → 4 → 3 → 2 → 5
